tests/test_non_empty_block.py

```diff
@@ -0,0 +1,322 @@
+    assert _test_value.pipe(non_empty_block.cons(-99)) == expected
+
+
+def test_non_empty_block_filter():
+    expected = Block([1, 2])
+    assert _test_value.filter(lambda i: i < 3) == expected
+    assert _test_value.pipe(non_empty_block.filter(lambda i: i < 3)) == expected
+
+
+def test_non_empty_block_fold():
+    initial_state = ''
+    expected = '1234'
+    assert _test_value.fold(lambda state, source: state + str(source), state=initial_state) == expected
+    assert _test_value.pipe(non_empty_block.fold(lambda state, source: state + str(source), state=initial_state)) == expected
+
+
+def test_non_empty_block_forall_true():
+    assert _test_value.forall(lambda i: i < 100)
+    assert _test_value.pipe(non_empty_block.forall(lambda i: i < 100))
+
+
+def test_non_empty_block_forall_false():
+    assert not _test_value.forall(lambda i: i < 4)
+    assert not _test_value.pipe(non_empty_block.forall(lambda i: i < 4))
+
+
+def test_non_empty_block_head():
+    assert _test_value.head() == 1
+    assert _test_value.pipe(non_empty_block.head) == 1
+
+
+def test_non_empty_block_indexed():
+    expected = NonEmptyBlock((0, 1), ((1, 2), (2, 3), (3, 4)))
+    assert _test_value.indexed() == expected
+    assert _test_value.pipe(non_empty_block.indexed) == expected
+
+
+def test_non_empty_block_item_first_item():
+    expected = Some(1)
+    assert _test_value.item(0) == expected
+    assert _test_value.pipe(non_empty_block.item(0)) == expected
+
+
+def test_non_empty_block_item_last_item():
+    expected = Some(4)
+    assert _test_value.item(3) == expected
+    assert _test_value.pipe(non_empty_block.item(3)) == expected
+
+
+def test_non_empty_block_item_out_of_range():
+    assert _test_value.item(99) is Nothing
+    assert _test_value.pipe(non_empty_block.item(99)) is Nothing
+
+
+def test_non_empty_block_map():
+    expected = NonEmptyBlock('X 1 X', ('X 2 X', 'X 3 X', 'X 4 X'))
+    assert _test_value.map(lambda i: f'X {i} X') == expected
+    assert _test_value.pipe(non_empty_block.map(lambda i: f'X {i} X')) == expected
+
+
+def test_non_empty_block_starmap():
+    mapped_value = _test_value.map(lambda i: (i, i * 10, i * 100))
+    expected = NonEmptyBlock('1 10 100', ('2 20 200', '3 30 300', '4 40 400'))
+    assert mapped_value.starmap(lambda a, b, c: f'{a} {b} {c}') == expected
+    assert mapped_value.pipe(non_empty_block.starmap(lambda a, b, c: f'{a} {b} {c}')) == expected
+
+
+# TODO: Not sure how to make this pass pyright
+#       Gives the following type of error:
+#         Type parameter "_TSource@NonEmptyBlock" is invariant, but "int" is not the same as
+#         "_TSourceSum@sum | Literal[0]"
+#def test_non_empty_block_sum():
+#    assert _test_value.sum() == 10
+#    assert _test_value.pipe(non_empty_block.sum) == 10
+#
+#
+#def test_non_empty_block_sum_by():
+#    assert _test_value.sum_by(lambda i: i * 10) == 100
+#    assert _test_value.pipe(non_empty_block.sum_by(lambda i: i * 10)) == 100
```

Comment:
Not sure why this doesn't work. I looked at `test_array.py#test_array_sum_works` which is similar and passes the type checker, and I can't see what I'm doing wrong here 🤷